sklearn/preprocessing/data.py

```diff
@@ -1766,138 +1781,257 @@ class OneHotEncoder(BaseEstimator, TransformerMixin):
     sklearn.preprocessing.LabelEncoder : encodes labels with values between 0
       and n_classes-1.
     """
-    def __init__(self, n_values="auto", categorical_features="all",
-                 dtype=np.float64, sparse=True, handle_unknown='error'):
-        self.n_values = n_values
+    def __init__(self, values='auto', categorical_features="all",
+                 n_values=None, dtype=np.float64, sparse=True,
+                 handle_unknown='error'):
+        self.values = values
         self.categorical_features = categorical_features
         self.dtype = dtype
         self.sparse = sparse
         self.handle_unknown = handle_unknown
+        self.n_values = n_values
 
     def fit(self, X, y=None):
-        """Fit OneHotEncoder to X.
+        """Fit the OneHotEncoder to X.
 
         Parameters
         ----------
         X : array-like, shape [n_samples, n_feature]
-            Input array of type int.
+            Array of ints or strings or both.
 
         Returns
         -------
         self
         """
-        self.fit_transform(X)
+        X = check_array(X, dtype=None, accept_sparse='csc', copy=False)
+        n_samples, n_features = X.shape
+        self.n_features_ = n_features
+
+        _apply_selected(X, self._fit, dtype=self.dtype, return_val=False,
+                        selected=self.categorical_features, copy=False)
+
+        # Record which columns of output data
+        # correspond to each column of input data
+        self.feature_index_range_ = np.zeros((n_features, 2), dtype=np.int)
+
+        if isinstance(self.categorical_features, six.string_types) and \
+           self.categorical_features == "all":
+            categorical = np.ones(n_features, dtype=bool)
+        else:
+            categorical = np.zeros(n_features, dtype=bool)
+            categorical[np.asarray(self.categorical_features)] = True
+
+        start, end = 0, 0
+        for i_cat, i_feat in enumerate(np.where(categorical)[0]):
+            if np.isscalar(self._values) and self._values == 'auto':
+                end = start + self.n_active_features_[i_cat]
+            else:
+                end = start + len(self._label_encoders[i_cat].classes_)
+            self.feature_index_range_[i_feat] = start, end
+            start = end
+        num_cat = np.sum(categorical)
+        non_cat_indices = np.arange(start, start + n_features - num_cat)
+        self.feature_index_range_[~categorical, 0] = non_cat_indices
+        self.feature_index_range_[~categorical, 1] = non_cat_indices + 1
+
+        # Record which column of input data corresponds
+        # to each column of output data
+        n_expanded_cols = end + n_features - num_cat
+        self.one_hot_feature_index_ = np.empty(n_expanded_cols, dtype=np.int)
```

Comment:
I think this can be filled with a one-liner (or something like it) using `np.repeat` and `np.diff`